blocks/blocks.php

```diff
@@ -47,11 +26,97 @@ function parse_blocks( $post_model ) {
 	// based on "edit_posts" cap. Since we want to serve blocks even to logged-out
 	// users -- and because we are parsing this content before returning it --
 	// we'll bypass the model and access the raw content directly.
-	$post = get_post( $post_model->ID );
+	$post = \get_post( $post_model->ID );
 
 	$is_gutenberg = \has_blocks( $post->post_content );
-	$blocks       = \parse_blocks( $post->post_content );
+	$raw_blocks   = \parse_blocks( $post->post_content );
+
+	return [
+		'blocks'      => process_content_blocks( $raw_blocks ),
+		'isGutenberg' => $is_gutenberg,
+		'version'     => $version,
+	];
+}
+
+/**
+ * Map the Gutenberg block attributes to the shape of BlockAttribute.
+ *
+ * @param  array $block Content block.
+ * @return array
+ */
+function get_content_block_attributes( $block ) {
+	if ( 'core/image' === $block['blockName'] ) {
+		$attachment_metadata = \wp_get_attachment_metadata( $block['attrs']['id'] );
+
+		$block['attrs']['src']            = \wp_get_attachment_url( $block['attrs']['id'] );
+		$block['attrs']['originalHeight'] = $attachment_metadata['height'];
+		$block['attrs']['originalWidth']  = $attachment_metadata['width'];
+		$block['attrs']['srcset']         = \wp_get_attachment_image_srcset( $block['attrs']['id'] );
+		$block['attrs']['alt']            = trim( strip_tags( \get_post_meta( $block['attrs']['id'], '_wp_attachment_image_alt', true ) ) );
+
+		// If width and height attributes aren't exposed, add the default ones
+		if ( ! isset( $block['attrs']['height'] ) ) {
+			$block['attrs']['height'] = $attachment_metadata['height'];
+		}
+
+		if ( ! isset( $block['attrs']['width'] ) ) {
+			$block['attrs']['width'] = $attachment_metadata['width'];
+		}
+	}
```

Comment:
I like that you're trying to include attributes that are saved in post content, and not parsed via the PHP parser here, but this seems a bit clunky to me to only enhance the image attributes and not others. Instead, maybe add a filter here so that additional attributes can be added to any block based on the block name and move the image attributes to that callback.

Ultimately, this is one of the annoying shortcomings of the PHP parser in WP, that it doesn't know how to parse out the attributes that are stored in markup and not on the block delimiters.

Reply:
Filter added in https://github.com/Automattic/vip-decoupled-bundle/pull/38/commits/bb0746170f339f769caf82f271d447df4aa02754

> Ultimately, this is one of the annoying shortcomings of the PHP parser in WP, that it doesn't know how to parse out the attributes that are stored in markup and not on the block delimiters.

_Sighs deeply_ 😭